### ISSUE TITLE
release-3.11: Running deploy_cluster.yml expects GlusterFS to be available before it is deployed

### ISSUE THREAD

**Issue by kareiva** [openshift/openshift-ansible#12002]
#### Description

When deploying OKD `release-3.11` on Centos7 with `deploy_cluster.yml`, the playbook `openshift-node/private/join.yml` is imported just before the component install[1]. 

The node join playbook induces a restart by importing the `restart.yml` playbook[2].

Restarting a node includes a task to check for GlusterFS health[3],[4], but at this point, the GlusterFS component installation has not yet been done.

[1] - https://github.com/openshift/openshift-ansible/blob/release-3.11/playbooks/deploy_cluster.yml#L12
[2] - https://github.com/openshift/openshift-ansible/blob/release-3.11/playbooks/openshift-node/private/join.yml#L102
[3] - https://github.com/openshift/openshift-ansible/blob/release-3.11/playbooks/openshift-node/private/restart.yml#L11
[4] - https://github.com/openshift/openshift-ansible/blob/release-3.11/roles/openshift_storage_glusterfs/tasks/cluster_health.yml

##### Version

* Ansible version per `ansible --version`

```
ansible 2.8.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/centos/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  7 2019, 00:51:29) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

* The output of `git describe`

```
openshift-ansible-3.11.154-1-2-g6298a66
```

##### Steps To Reproduce
1. Setup clean Centos-7.7 nodes and adjust for OKD cluster deployment
2. Configure the host inventory with below non-default glusterfs variables:
```
openshift_hosted_registry_storage_kind=glusterfs
openshift_storage_glusterfs_health_timeout=300

[glusterfs]
node-[0:2].okd glusterfs_devices='[ "/dev/vdc" ]'
```
3. Run the `prerequisites.yml` playbook
4. Run the `deploy_cluster.yml`

##### Expected Results
OKD cluster is being installed with glusterfs storage.

##### Observed Results
Node-join phase fails with GlusterFS health check before GlusterFS is installed:

```
TASK [openshift_storage_glusterfs : GlusterFS cluster health time warning] 
ok: [master-0.okd] => {
    "msg": "Depending on the number of volumes in the cluster and their size, this might take few minutes to a couple of hours"
}

TASK [openshift_storage_glusterfs : Check for GlusterFS cluster health] 
FAILED - RETRYING: Check for GlusterFS cluster health (30 retries left).
FAILED - RETRYING: Check for GlusterFS cluster health (1 retries left).
fatal: [master-0.okd -> master-0.okd]: FAILED! => {"attempts": 30, "changed": false, "msg": "Unable to find pods: ", "state": "unknown"}
```

##### Additional Information

```
$ cat /etc/redhat-release
CentOS Linux release 7.7.1908 (Core)

```

**Comment by CygnusHyoga** [comment]
I've the same problem, with sensible 2.8.5 and too with CentOS Linux release 7.7.1908 (Core). Here is my debug log of the task:

Using module file /Users/kulo4/openshift-ansible/roles/lib_utils/library/glusterfs_check_containerized.py
Pipelining is enabled.
<minion-master.buenhost.com.ar> ESTABLISH SSH CONNECTION FOR USER: root
<minion-master.buenhost.com.ar> SSH: EXEC sshpass -d103 ssh -vvv -o ControlMaster=auto -o ControlPersist=600s -o StrictHostKeyChecking=no -o 'User="root"' -o ConnectTimeout=30 -o ControlPath=/Users/kulo4/.ansible/cp/%h-%r minion-master.buenhost.com.ar '/bin/sh -c '"'"'/usr/bin/python && sleep 0'"'"''
<minion-master.buenhost.com.ar> (1, b'\n{"msg": "Unable to find pods: ", "failed": true, "state": "unknown", "changed": false, "invocation": {"module_args": {"target_nodes": null, "exclude_node": "minion-master.buenhost.com.ar", "oc_namespace": "storage", "cluster_name": "storage", "oc_bin": "oc", "check_bricks": false, "oc_conf": "/etc/origin/master/admin.kubeconfig"}}}\n', b'OpenSSH_7.9p1, LibreSSL 2.7.3\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 48: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 68858\r\ndebug3: mux_client_request_session: session request sent\r\n/etc/profile.d/lang.sh: line 19: warning: setlocale: LC_CTYPE: cannot change locale (UTF-8): No such file or directory\nNo resources found.\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\n')
<minion-master.buenhost.com.ar> Failed to connect to the host via ssh: OpenSSH_7.9p1, LibreSSL 2.7.3
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 48: Applying options for *
debug1: auto-mux: Trying existing master
debug2: fd 3 setting O_NONBLOCK
debug2: mux_client_hello_exchange: master version 4
debug3: mux_client_forwards: request forwardings: 0 local, 0 remote
debug3: mux_client_request_session: entering
debug3: mux_client_request_alive: entering
debug3: mux_client_request_alive: done pid = 68858
debug3: mux_client_request_session: session request sent
/etc/profile.d/lang.sh: line 19: warning: setlocale: LC_CTYPE: cannot change locale (UTF-8): No such file or directory
No resources found.
debug3: mux_client_read_packet: read header failed: Broken pipe
debug2: Received exit status from master 1
fatal: [minion-master.buenhost.com.ar -> minion-master.buenhost.com.ar]: FAILED! => {
    "attempts": 3,
    "changed": false,
    "invocation": {
        "module_args": {
            "check_bricks": false,
            "cluster_name": "storage",
            "exclude_node": "minion-master.buenhost.com.ar",
            "oc_bin": "oc",
            "oc_conf": "/etc/origin/master/admin.kubeconfig",
            "oc_namespace": "storage",
            "target_nodes": null
        }
    },
    "msg": "Unable to find pods: ",
    "state": "unknown"
}

**Comment by awasthan** [comment]
I am also facing same error on RHEL 7.6 with ansible 2.6.12

**Comment by s-abe0** [comment]
Same issue here running deployment on CentOS in AWS. My current workaround is to run glusterfs install script after the openshift install has completed: 

run `openshift-ansible/playbooks/deploy_cluster.yml` first, then run `openshift-ansible/playbooks/openshift-glusterfs/config.yml`.

Of course you will need two separate inventory files, one with glusterfs settings commented out, another with glusterfs settings enabled that will be used by glusterfs config playbook.

**Comment by nacerwadghiri** [comment]
Regarding this issue, the best solution would be to delete in openshift-ansible/playbooks/openshift-node/private/restart.yml the following parts :

  - name: Check Gluster Health Before Restart
    include_role:
      name: openshift_storage_glusterfs
      tasks_from: check_cluster_health.yml

  - name: Check Gluster Health After Restart
    include_role:
      name: openshift_storage_glusterfs
      tasks_from: check_cluster_health.yml

Another solution if you already ran the  openshift-ansible/playbooks/deploy_cluster.yml playbook would be to modify it as below and reran it. It will continue the installation as it is done in one-shot but without restarting kubelet and docker. 

---
- import_playbook: init/main.yml
  vars:
    l_prereq_check_hosts: "oo_nodes_to_config"

- import_playbook: common/private/components.yml

**Comment by stankeB** [comment]
Is also an issue in OSE https://bugzilla.redhat.com/show_bug.cgi?id=1777499&GoAheadAndLogIn=1.

Best way would be to add a condition, that checks in which phase installation is, because the playbook is also used during upgrade, were checking gluster state is totally right.

**Comment by kevchu3** [comment]
I'm also running into the same issue on OCP v3.11.154 and RHEL 7.7.  I ended up relabeling my `glusterfs` nodes as `appnodes` and then picking up the install after it passed the `openshift-node/join.yml` playbook.

**Comment by stankeB** [comment]
Give it a new try today and it is working now.

Package versions:
openshift-ansible-3.11.157-1.git.0.10b76ed.el7.noarch
openshift-ansible-playbooks-3.11.157-1.git.0.10b76ed.el7.noarch
openshift-ansible-roles-3.11.157-1.git.0.10b76ed.el7.noarch

**Comment by sdodson** [comment]
/close
Fixed via #11987
